ENIGMAsystem/SHELL/Graphics_Systems/General/GSmodel.cpp

```diff
@@ -15,15 +17,651 @@
+    }
+  }
+  if (!next.empty())
+    result.push_back(atof(next.c_str()));
+  return result;
+}
+
+//obj model parsing functions
+void string_parse(string *s) {
+  size_t spaces = 0;
+  bool trimmed = false;
+  bool checknormal = false;
+  for (unsigned int i = 0; i < s->size() ; i++)
+  {
+    //comment
+    if ((*s)[i] == '#')
+    {
+      s->erase(i, s->length() - i);
+      break;
+    }
+    else if((*s)[i] == ' ')
+    {
+      if (!trimmed)
+      {
+        s->erase(i,1);
+        i--;
+      }
+      else
+      {
+        if (spaces >= 1)
+        {
+          s->erase(i,1);
+          i--;
+        }
+        spaces++;
+      }
+    }
+    else
+    {
+      if((*s)[i] == '/')
+      {
+        (*s)[i] = ' ';
+        if(checknormal)
+        {
+          s->erase(i, 1);
+          checknormal = false;
+        }
+        else
+          checknormal = true;
+      }
+      else
+        checknormal = false;
+      spaces = 0;
+      trimmed = true;
+    }
+  }
+  //end trim
+  if (s->size() > 0) {
+    if ((*s)[s->size()-1] == ' ')
+    {
+      s->erase(s->size()-1, 1);
+    }
+  }
+}
+
+} // namespace enigma
+
 namespace enigma_user {
 
+int d3d_model_create(int type) {
+  int id = enigma::models.size();
+  enigma::Model* model = new enigma::Model(type);
+  model->vertex_buffer = vertex_create_buffer();
+  enigma::models.push_back(model);
+  return id;
+}
+
+void d3d_model_destroy(int id) {
+  const enigma::Model* model = enigma::models[id];
+  vertex_delete_buffer(model->vertex_buffer);
+
+  delete enigma::models[id];
+  enigma::models[id] = nullptr;
+}
+
+bool d3d_model_exists(int id) {
+  return (id >= 0 && (unsigned)id < enigma::models.size() && enigma::models[id] != nullptr);
+}
+
+void d3d_model_clear(int id) {
+  enigma::Model* model = enigma::models[id];
+  vertex_clear(model->vertex_buffer);
+  model->primitives.clear();
+  model->current_primitive = 0;
+  model->vertex_started = false;
+}
+
+void d3d_model_draw(int id) {
+  enigma::Model* model = enigma::models[id];
+  if (model->vertex_started) {
+    model->vertex_started = false;
+    vertex_end(model->vertex_buffer);
+
+    // freeze the model if it's static to indicate to the driver that
+    // we don't intend on updating this sucker so it will draw faster
+    if (model->type == model_static) {
+      vertex_freeze(model->vertex_buffer, false);
+    } else if (model->type == model_dynamic) {
+      vertex_freeze(model->vertex_buffer, true);
+    }
+    // model_stream type is never frozen because it means the user
+    // will be updating and specifying new primitives every frame
+  }
+  for (auto primitive : model->primitives) {
+    vertex_set_format(model->vertex_buffer, primitive.format);
+    vertex_submit(
+      model->vertex_buffer,
+      primitive.type,
+      primitive.vertex_start,
+      primitive.vertex_count
+    );
+  }
+}
+
+void d3d_model_draw(int id, gs_scalar x, gs_scalar y, gs_scalar z) {
+  d3d_transform_add_translation(x, y, z);
```

Comment:
GL1 may perform better if you use `d3d_transform_push_translation(x, y, z)` / `d3d_transform_pop()`. Don't feel obligated to try this before submitting this change.

Reply:
That will break Direct3D9 where I haven't written those functions yet, we can look at doing that later and benchmark it. 👍